File: src/servent/SimpleServentListener.java
package servent;

import java.io.IOException;
import java.net.ServerSocket;
import java.net.Socket;
import java.net.SocketTimeoutException;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

import app.AppConfig;
import app.Cancellable;
import app.snapshot_bitcake.LYSnapshotResult;
import app.snapshot_bitcake.LaiYangBitcakeManager;
import app.snapshot_bitcake.Pair;
import app.snapshot_bitcake.SnapshotCollector;
import servent.handler.MessageHandler;
import servent.handler.NullHandler;
import servent.handler.TransactionHandler;
import servent.handler.snapshot.LYMarkerHandler;
import servent.handler.snapshot.LYTellHandler;
import servent.handler.snapshot.RoundHandler;
import servent.message.Message;
import servent.message.MessageType;
import servent.message.snapshot.LYMarkerMessage;
import servent.message.snapshot.LYTellMessage;
import servent.message.util.MessageUtil;

public class SimpleServentListener implements Runnable, Cancellable {

	private volatile boolean working = true;

	private SnapshotCollector snapshotCollector;

	public SimpleServentListener(SnapshotCollector snapshotCollector) {
		this.snapshotCollector = snapshotCollector;
	}

	/*
	 * Thread pool for executing the handlers. Each client will get it's own handler thread.
	 */
	private final ExecutorService threadPool = Executors.newWorkStealingPool();

	private List<Message> redMessages = new ArrayList<>();

	@Override
	public void run() {
		ServerSocket listenerSocket = null;
		try {
			listenerSocket = new ServerSocket(AppConfig.myServentInfo.getListenerPort(), 100);
			/*
			 * If there is no connection after 1s, wake up and see if we should terminate.
			 */
			listenerSocket.setSoTimeout(1000);
		} catch (IOException e) {
			AppConfig.timestampedErrorPrint("Couldn't open listener socket on: " + AppConfig.myServentInfo.getListenerPort());
			System.exit(0);
		}


		while (working) {
			try {
				Message clientMessage;

				/*
				 * Lai-Yang stuff. Process any red messages we got before we got the marker.
				 * The marker contains the collector id, so we need to process that as our first
				 * red message. 
				 */
				//if (AppConfig.isWhite.get() == false && redMessages.size() > 0) {
				//	clientMessage = redMessages.remove(0);
				//} else {
				/*
					 * This blocks for up to 1s, after which SocketTimeoutException is thrown.
					 */
				Socket clientSocket = listenerSocket.accept();

				//GOT A MESSAGE! <3
				clientMessage = MessageUtil.readMessage(clientSocket);
				//}
				//synchronized (AppConfig.colorLock) {
				//if (clientMessage.isWhite() == false && AppConfig.isWhite.get()) {
				/*
						 * If the message is red, we are white, and the message isn't a marker,
						 * then store it. We will get the marker soon, and then we will process
						 * this message. The point is, we need the marker to know who to send
						 * our info to, so this is the simplest way to work around that.
						 */
				//if (clientMessage.getMessageType() != MessageType.LY_MARKER) {
				//	redMessages.add(clientMessage);
				//	continue;
				//} else {
				if (clientMessage.getMessageType() == MessageType.LY_MARKER) {
					System.out.println("RECEIVED MARKER!");
					if (((LYMarkerMessage)clientMessage).getLyTellMessage() != null) {
						System.out.println("RECEIVED RESULTS FROM CHILD!");
						System.out.print("IVE RECEIVED IN LISTENER THIS: ");
						for (LYSnapshotResult result : ((LYMarkerMessage)clientMessage).getLyTellMessage().getLYSnapshotResult()){
							System.out.println(result.getServentId());
						}
					}
					LaiYangBitcakeManager lyFinancialManager =
							(LaiYangBitcakeManager)snapshotCollector.getBitcakeManager();
					lyFinancialManager.markerEvent(
							new Pair(((LYMarkerMessage)clientMessage).getCollectorId() , ((LYMarkerMessage)clientMessage).getSnapshotNumber()), clientMessage.getOriginalSenderInfo().getId(), snapshotCollector, ((LYMarkerMessage)clientMessage).getLyTellMessage());
				}
				//}
				//}

				MessageHandler messageHandler = new NullHandler(clientMessage);

				/*
				 * Each message type has it's own handler.
				 * If we can get away with stateless handlers, we will,
				 * because that way is much simpler and less error prone.
				 */
				switch (clientMessage.getMessageType()) {
					case TRANSACTION:
						messageHandler = new TransactionHandler(clientMessage, snapshotCollector.getBitcakeManager());
						break;
					case LY_MARKER:
						messageHandler = new LYMarkerHandler();
						break;
					case ROUND:
						messageHandler = new RoundHandler(clientMessage, snapshotCollector);
				}

				threadPool.submit(messageHandler);
			} catch (SocketTimeoutException timeoutEx) {
				//Uncomment the next line to see that we are waking up every second.
//				AppConfig.timedStandardPrint("Waiting...");
			} catch (IOException e) {
				e.printStackTrace();
			}
		}
	}

	@Override
	public void stop() {
		this.working = false;
	}

}

File: Snapshot_GO/servent/simple_servent_listener.go
package servent

import (
	"Snapshot/app"
	"Snapshot/app/snapshotbitcake"
	"Snapshot/helper"
	"Snapshot/servent/handler"
	handlersnapshot "Snapshot/servent/handler/snapshot"
	"Snapshot/servent/message"
	"log"
	"net"
	"strconv"
	"sync"
	"time"
)

type SimpleServentListener struct {
	working           bool
	snapshotCollector snapshotbitcake.SnapshotCollector
	redMessages       []message.Message
	threadPool        *helper.Pool
}

func ConstructSimpleServentListener(snapshotCollector snapshotbitcake.SnapshotCollector) *SimpleServentListener {
	threadPool := helper.InitiatePool()
	return &SimpleServentListener{true, snapshotCollector, nil, threadPool}
}

func (simpleServentListener SimpleServentListener) Run(wgListener *sync.WaitGroup) {
	listener, err := net.Listen("tcp", app.MyServentInfo.Ip+":"+strconv.Itoa(app.MyServentInfo.Port))
	app.TimeStampPrint("Now waiting for connection on: " + strconv.Itoa(app.MyServentInfo.Port))
	wgListener.Done()

	if err != nil {
		log.Println(err)
		app.ExitWithErrorMsg("ERROR ESTABLISHING SERVENT LISTENER ON PORT: " + strconv.Itoa(app.MyServentInfo.Port))
	}
	defer listener.Close()

	for {
		// If working is set to false, drop out.
		if !simpleServentListener.working {
			break
		}

		var clientMessage message.Message = nil
		var connection net.Conn

		if !app.IsWhite.Load() && len(simpleServentListener.redMessages) > 0 {
			// Remove first element.
			clientMessage = simpleServentListener.redMessages[0]
			simpleServentListener.redMessages = simpleServentListener.redMessages[1:]
		} else {
			// Wait max for 1s for listener to accept.
			listener.(*net.TCPListener).SetDeadline(time.Now().Add(time.Second))
			// Listener will wait max 1s, and then repeat the for loop.
			connection, err = listener.Accept()

			if err != nil {
				log.Println(err)
				// The connection didn't receive any calls, so
				// we go back to the top to check if working == true.
				continue
			}
			app.TimeStampPrint("Received connection on : " + strconv.Itoa(app.MyServentInfo.Port))

		}

		go func() {
			if clientMessage != nil {
				clientMessage = snapshotbitcake.ReadMessage(connection)
			}

			if clientMsg, ok := clientMessage.(*message.BasicMessage); ok {
				app.ColorLock.Lock()

				if !clientMsg.White && app.IsWhite.Load() {
					/*
					* If the message is red, we are white, and the message isn't a marker,
					* then store it. We will get the marker soon, and then we will process
					* this message. The point is, we need the marker to know who to send
					* our info to, so this is the simplest way to work around that.
					 */
					if clientMsg.MessageType != "LY_MARKER" {
						simpleServentListener.redMessages = append(simpleServentListener.redMessages, clientMsg)
						return
					} else {
						if lybm, ok := simpleServentListener.snapshotCollector.GetBitcakeManager().(*snapshotbitcake.LaiYangBitcakeManager); ok {
							tmp, err := strconv.Atoi(clientMsg.MessageText)

							if err != nil {
								app.TimeStampErrorPrint("Error parsing integer: " + clientMsg.MessageText)
							}

							lybm.MarkerEvent(tmp, simpleServentListener.snapshotCollector)
						} else {
							app.TimeStampErrorPrint("Couldnt get bitcake manager to LaiYangBitcakeManager.")
						}
					}

					app.ColorLock.Unlock()

					var messageHandler handler.MessageHandler = handler.NullHandler{ClientMessage: clientMsg}

					switch clientMsg.MessageType {
					case "TRANSACTION":
						messageHandler = snapshotbitcake.TransactionalHandler{ClientMessage: clientMsg, BitcakeManager: simpleServentListener.snapshotCollector.GetBitcakeManager()}
					case "LY_MARKER":
						messageHandler = handlersnapshot.LYMarkerHandler{}
					case "LY_TELL":
						messageHandler = handlersnapshot.LYTellHandler{ClientMessage: clientMsg, SnapshotCollector: simpleServentListener.snapshotCollector}
					}

					// Add the message handler to the thread pool.
					simpleServentListener.threadPool.JobQueue <- helper.Job{Handler: messageHandler}
				}
			} else {
				app.TimeStampErrorPrint("Message wasn't a BasicMessage!")

			}
		}()
	}
}

func (simpleServentListener *SimpleServentListener) Stop() {
	(*simpleServentListener).working = false
}
